File: 0201-bitwise-and-of-numbers-range/0201-bitwise-and-of-numbers-range.rb
# @param {Integer} left
# @param {Integer} right
# @return {Integer}
POWERS, p = [0], 1
while p <= 2147483647
  POWERS << p
  p *= 2
end

def range_bitwise_and(left, right)
  return 0 if left == 0
  l_idx, r_idx, i = -1, -1, 0
  while i < POWERS.length
    l_idx += 1 if left >= POWERS[i]
    r_idx += 1 if right >= POWERS[i]
    i += 1
  end
  return 0 if l_idx != r_idx
  base = POWERS[l_idx]
  base+range_bitwise_and(left-base, right-base)
end

File: 0201-bitwise-and-of-numbers-range/0201-bitwise-and-of-numbers-range.js
/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */

const powers = [0];
let p = 1;
while (p <= 2147483647) {
  powers.push(p);
  p *= 2;
}

var rangeBitwiseAnd = function(left, right) {
  if (left === 0) return 0;
  let [lIdx, rIdx, i] = [-1, -1, 0];
  while (i < powers.length) {
    if (left >= powers[i]) lIdx++;
    if (right >= powers[i]) rIdx++;
    i++;
  }
  if (lIdx !== rIdx) return 0;
  return powers[lIdx] + rangeBitwiseAnd(left - powers[lIdx], right - powers[lIdx]);
}